### PR TITLE
Enum support for if else to switch cleanup

### PR DESCRIPTION
**Pull request by carstenartur**
## What it does
Adds support for enum to the cleanup in charge for refactoring chains of "if else" to switch.

## How to test
Run the cleanup on code with if else chains or add junit test cases.

## Author checklist

- [ ] I have thoroughly tested my changes
- [x] The change is following the [coding conventions](https://wiki.eclipse.org/Platform/How_to_Contribute#Coding_Conventions)
- [x] I have signed the [Eclipse Contributor Agreement (ECA)](https://www.eclipse.org/legal/ECA.php)

### CODE REVIEW

**Comment by carstenartur** [comment]
Hi @jjohnstn, can you have a look? The build error might be unrelated, not sure.

File: org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/fix/SwitchFixCore.java [hunk 1]
```diff
@@ -237,6 +238,9 @@ private List<SwitchCaseSection> filterDuplicateCaseValues(final List<SwitchCaseS
 
 					for (Expression expression : sourceCase.literalExpressions) {
 						Object constantValue= expression.resolveConstantExpressionValue();
+						if(constantValue == null && expression.resolveTypeBinding().isEnum()) {
```
Comment (jjohnstn) [inline org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/fix/SwitchFixCore.java:241]: You always have to check for null with binding resolution.  Sometimes it is just the fact that there is an error somewhere else in the file.

File: org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/fix/SwitchFixCore.java [hunk 2]
```diff
@@ -323,13 +327,19 @@ private Variable extractVariableAndValuesFromInfixExpression(final InfixExpressi
...
-						&& ASTNodes.hasType(variable, char.class.getCanonicalName(),
-								byte.class.getCanonicalName(), short.class.getCanonicalName(),
-								int.class.getCanonicalName())
-						&& constant.resolveTypeBinding() != null
-						&& constant.resolveTypeBinding().isPrimitive()
-						&& constant.resolveConstantExpressionValue() != null) {
+				if (!(variable instanceof Name || variable instanceof FieldAccess || variable instanceof SuperFieldAccess))return null;
+
+				ITypeBinding variabletypeBinding= variable.resolveTypeBinding();
+				boolean isVariableEnum= variabletypeBinding.isEnum();
```
Comment (jjohnstn) [inline org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/fix/SwitchFixCore.java:333]: Need to add null check for the type binding.

File: org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/fix/SwitchFixCore.java [hunk 3]
```diff
@@ -323,13 +327,19 @@ private Variable extractVariableAndValuesFromInfixExpression(final InfixExpressi
...
+				boolean hasType= ASTNodes.hasType(variable, char.class.getCanonicalName(),
+						byte.class.getCanonicalName(), short.class.getCanonicalName(),
+						int.class.getCanonicalName()) || isVariableEnum;
+
+				ITypeBinding constanttypeBinding= constant.resolveTypeBinding();
+				if (hasType
+						&& constanttypeBinding != null
+						&& (constanttypeBinding.isPrimitive() || isVariableEnum)
+						&& (constant.resolveConstantExpressionValue() != null|| isVariableEnum)) {
```
Comment (jjohnstn) [inline org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/fix/SwitchFixCore.java:342]: formatting nit.  Add blank between null and ||

File: org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/fix/SwitchFixCore.java [hunk 4]
```diff
@@ -323,13 +327,19 @@ private Variable extractVariableAndValuesFromInfixExpression(final InfixExpressi
...
 
 			private Variable extractVariableWithConstantValue(final Expression variable, final Expression constant) {
-				if ((variable instanceof Name || variable instanceof FieldAccess || variable instanceof SuperFieldAccess)
-						&& ASTNodes.hasType(variable, char.class.getCanonicalName(),
-								byte.class.getCanonicalName(), short.class.getCanonicalName(),
-								int.class.getCanonicalName())
-						&& constant.resolveTypeBinding() != null
-						&& constant.resolveTypeBinding().isPrimitive()
-						&& constant.resolveConstantExpressionValue() != null) {
+				if (!(variable instanceof Name || variable instanceof FieldAccess || variable instanceof SuperFieldAccess))return null;
```
Comment (jjohnstn) [inline org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/fix/SwitchFixCore.java:330]: Minor nit.  Please put return on new line.

**Review by jjohnstn — CHANGES_REQUESTED** [review]

**Review by jjohnstn — APPROVED** [review]

**Comment by jjohnstn** [comment]
Thanks @carstenartur

**Pull request merged**